File: haskell/bob/Iteration3/src/Bob.hs
module Bob (responseFor) where

import Data.Char
import Data.List

responseFor :: String -> String
responseFor phrase
  | isSilence phrase          = "Fine. Be that way!"
  | isShoutingQuestion phrase = "Calm down, I know what I'm doing!"
  | isQuestion phrase         = "Sure."
  | isShouting phrase         = "Whoa, chill out!"
  | otherwise                 = "Whatever."

isShoutingQuestion :: String -> Bool
isShoutingQuestion = (&&) <$> isShouting <*> isQuestion

isQuestion :: String -> Bool
isQuestion = (== '?') . last . dropWhileEnd isSpace

isShouting :: String -> Bool
isShouting = (&&) <$> any isUpper <*> (all isUpper . filter isAlpha)

isSilence :: String -> Bool
isSilence = (||) <$> null <*> all isSpace

File: csharp/bob/Iteration3/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    public static string Response(string statement)
    {
        if (IsSilent(statement)) return "Fine. Be that way!";

        var isShouting = IsShouting(statement);
        var isQuestion = IsQuestion(statement);
        
        if (isQuestion && isShouting) return "Calm down, I know what I'm doing!";
        if (isQuestion) return "Sure.";
        if (isShouting) return "Whoa, chill out!";

        return "Whatever.";
    }

    private static bool IsQuestion(string statement) =>
        statement.Length > 0 && statement.Trim()[^1] == '?';

    private static bool IsShouting(string statement) =>
        statement.Any(Char.IsLetter) && statement.ToUpper() == statement;

    private static bool IsSilent(string statement) =>
        string.IsNullOrWhiteSpace(statement) || statement.All(Char.IsControl);
}
